File: tools/assets/canopy-visibility.py
# SPDX-License-Identifier: GPL-3.0-or-later
"""Bake raw alpha-aware diffuse accessibility from declared original oak sources."""
import argparse, hashlib, json, math, pathlib, sys, time
import bpy, numpy as np
from mathutils import Vector, geometry
from mathutils.bvhtree import BVHTree

parser=argparse.ArgumentParser()
parser.add_argument('--project-root',type=pathlib.Path,required=True)
parser.add_argument('--models',default='tree,roadside-oak')
parser.add_argument('--blender-version',default='5.2.1')
parser.add_argument('--output-root',type=pathlib.Path,required=True)
parser.add_argument('--samples',type=int,default=256)
args=parser.parse_args(sys.argv[sys.argv.index('--')+1:])
assert args.blender_version=='5.2.1', 'Retained oak source and visibility bake require verified Blender5.2.1'
assert bpy.app.version[:3]==tuple(map(int,args.blender_version.split('.')))
assert args.samples==256
ROOT=args.project_root.resolve();OUT=args.output_root.resolve()/'assets/source/canopy';OUT.mkdir(parents=True,exist_ok=True)
BASE=ROOT/'assets/source/canopy-base'
models=args.models.split(',');assert models and len(models)==len(set(models)) and all(n in ['tree','roadside-oak'] for n in models)
sha=lambda p:hashlib.sha256(pathlib.Path(p).read_bytes()).hexdigest()
base=json.loads((BASE/'base.json').read_text());assert base['version']==1 and base['blenderVersion']==args.blender_version
for name in models:
 for kind in ['blend','glb']:
  entry=base['models'][name][kind];assert entry['path']==f'assets/source/canopy-base/{name}.{kind}';p=ROOT/entry['path'];assert sha(p)==entry['sha256'] and p.stat().st_size==entry['bytes']
inputs={str(p.relative_to(ROOT)):sha(p) for p in [BASE/'base.json',*[BASE/f'{name}.{kind}' for name in models for kind in ['blend','glb']],ROOT/'assets/textures/arroyo-foliage.webp',pathlib.Path(__file__).resolve()]}
directions=[]
for i in range(args.samples):
 z=1-2*(i+.5)/args.samples;a=i*2.399963229728653;r=math.sqrt(1-z*z)
 directions.append(Vector((math.cos(a)*r,math.sin(a)*r,z)))
results=[]
for name in models:
 started=time.time();bpy.ops.wm.open_mainfile(filepath=str(BASE/f'{name}.blend'),use_scripts=False)
 bpy.context.preferences.filepaths.save_version=0
 image=bpy.data.images.load(str(ROOT/'assets/textures/arroyo-foliage.webp'),check_existing=False)
 width,height=image.size;pixels=np.empty(width*height*4,dtype=np.float32);image.pixels.foreach_get(pixels)
 alpha=pixels.reshape(height,width,4)[:,:,3]
 ys,xs=np.nonzero(alpha>=115/255)
 opaque_uv=np.column_stack(((xs+.5)/width,(ys+.5)/height))
 nearest={}
 for u in [0,.5,1]:
  for v in [0,.5,1]:
   nearest[(u,v)]=tuple(opaque_uv[np.argmin(np.sum((opaque_uv-np.array((u,v)))**2,axis=1))])
 objects=[o for o in bpy.context.scene.objects if o.type=='MESH'];bpy.context.view_layer.update()
 triangles=[];bvh_vertices=[];bvh_indices=[];leaf_data=[]
 for o in objects:
  m=o.data;m.calc_loop_triangles();normal_matrix=o.matrix_world.to_3x3().inverted().transposed()
  is_leaf=all(mat.name.startswith('foliage') for mat in m.materials)
  uv=m.uv_layers.active
  for tri in m.loop_triangles:
   points=[o.matrix_world@m.vertices[i].co for i in tri.vertices]
   tex=[Vector((*uv.data[i].uv,0)) for i in tri.loops] if is_leaf else None
   index=len(triangles);triangles.append((points,tex,is_leaf,o,tuple(tri.vertices)))
   begin=len(bvh_vertices);bvh_vertices.extend(points);bvh_indices.append((begin,begin+1,begin+2))
  if is_leaf:leaf_data.append((o,normal_matrix))
 bvh=BVHTree.FromPolygons(bvh_vertices,bvh_indices,all_triangles=True)
 ray_count=0;transparent_steps=0
 def visible(point,direction):
  global ray_count,transparent_steps
  ray_count+=1;origin=point+direction*.005;remaining=32.
  for step in range(512):
   hit,_,index,distance=bvh.ray_cast(origin,direction,remaining)
   if hit is None:return 1.
   points,tex,is_leaf,_,_=triangles[index]
   if not is_leaf:return 0.
   mapped=geometry.barycentric_transform(hit,*points,*tex)
   x=min(width-1,max(0,int(mapped.x*width)));y=min(height-1,max(0,int(mapped.y*height)))
   if alpha[y,x]>=115/255:return 0.
   transparent_steps+=1;remaining-=distance+.00002
   if remaining<=0:return 1.
   origin=hit+direction*.00002
  raise RuntimeError('Transparent ray traversal exceeded its guarded bound')
 cache={};sample_cache={};displacements=[];values=[];components_total=0
 for o,normal_matrix in leaf_data:
  m=o.data;adj=[set() for _ in m.vertices]
  for e in m.edges:a,b=e.vertices;adj[a].add(b);adj[b].add(a)
  component={};groups=[]
  for i in range(len(m.vertices)):
   if i in component:continue
   pending=[i];group=[];ci=len(groups)
   while pending:
    k=pending.pop()
    if k in component:continue
    component[k]=ci;group.append(k);pending.extend(adj[k])
   assert len(group)==9, (o.name,len(group));groups.append(group)
  components_total+=len(groups)
  local_tris={ci:[] for ci in range(len(groups))}
  for points,tex,is_leaf,owner,indices in triangles:
   if owner==o:local_tris[component[indices[0]]].append((points,tex))
  attribute=m.attributes.new('_CANOPY_VISIBILITY','FLOAT2','CORNER')
  uv=m.uv_layers.active
  for loop in m.loops:
   point=o.matrix_world@m.vertices[loop.vertex_index].co
   normal=(normal_matrix@m.corner_normals[loop.index].vector).normalized()
   coord=tuple(round(float(x),7) for x in uv.data[loop.index].uv)
   assert coord in nearest,coord
   equivalence=tuple(round(float(x),8) for x in (*point,*normal,*coord))
   if equivalence not in cache:
    ck=(o.name,component[loop.vertex_index],coord)
    if ck not in sample_cache:
     target=Vector((*nearest[coord],0));sample=None
     for points,tex in local_tris[component[loop.vertex_index]]:
      if geometry.intersect_point_tri_2d(target,tex[0],tex[1],tex[2]):
       sample=geometry.barycentric_transform(target,*tex,*points);break
     assert sample is not None,ck
     displacements.append((sample-point).length)
     openness=np.array([visible(sample,d) for d in directions],dtype=np.float64)
     sample_cache[ck]=openness
    weights=np.array([normal.dot(d) for d in directions],dtype=np.float64)
    forward=np.maximum(weights,0);backward=np.maximum(-weights,0);openness=sample_cache[ck]
    value=(float(openness@forward/forward.sum()),float(openness@backward/backward.sum()))
    assert all(math.isfinite(v) and 0<=v<=1 for v in value),value
    cache[equivalence]=value;values.append(value)
   attribute.data[loop.index].vector=cache[equivalence]
  print(json.dumps({'model':name,'mesh':o.name,'groups':len(groups),'cachedOpaqueOrigins':len(sample_cache),'elapsedSeconds':time.time()-started}),flush=True)
 assert components_total==630
 source=OUT/f'{name}.blend';glb=OUT/f'{name}.glb'
 bpy.ops.wm.save_as_mainfile(filepath=str(source),compress=True)
 bpy.ops.export_scene.gltf(filepath=str(glb),export_format='GLB',export_animations=True,export_animation_mode='ACTIONS',export_force_sampling=True,export_materials='EXPORT',export_yup=True,export_attributes=True)
 result={'model':name,'samples':args.samples,'sourceSha256':sha(source),'glbSha256':sha(glb),'sprays':components_total,'equivalenceClasses':len(cache),'opaqueOrigins':len(sample_cache),'rays':ray_count,'transparentSteps':transparent_steps,'atlasSize':[width,height],'nearestOpaqueUv':{str(k):v for k,v in nearest.items()},'originDisplacementMetres':{'maximum':max(displacements),'p90':float(np.quantile(displacements,.9)),'mean':float(np.mean(displacements))},'rawAccessibility':{'minimum':float(np.min(values)),'maximum':float(np.max(values)),'mean':float(np.mean(values))},'elapsedSeconds':time.time()-started}
 results.append(result);print(json.dumps(result),flush=True)
for p,h in inputs.items():assert sha(ROOT/p)==h,p
(OUT/'native.json').write_text(json.dumps({'inputs':inputs,'blender':bpy.app.version_string,'rows':results,'method':'Original retained native meshes, unchanged folded geometry/UV/normals/materials. Every ray origin lies on an actual original-alpha opaque texel in the same connected spray. Raw cosine-weighted two-sided diffuse accessibility, no strength/gain/floor/clamp;512 transparent-hit traversal guard raises on exhaustion. Geometry and all original GLB fields require separate byte-exact export verification before use.'},indent=2)+'\n')

File: tools/assets/canopy-visibility.mjs
// SPDX-License-Identifier: GPL-3.0-or-later
import assert from 'node:assert/strict';
import {readFileSync, writeFileSync, mkdirSync, copyFileSync} from 'node:fs';
import {createHash} from 'node:crypto';
import path from 'node:path';
import {GLTFLoader} from 'three/addons/loaders/GLTFLoader.js';

const sha = bytes => createHash('sha256').update(bytes).digest('hex');
const identify = file => { const bytes=readFileSync(file); return {bytes:bytes.length,sha256:sha(bytes)}; };
const bytesOf = attribute => Buffer.from(attribute.array.buffer,attribute.array.byteOffset,attribute.array.byteLength);
const field = attribute => ({itemSize:attribute.itemSize,count:attribute.count,normalized:attribute.normalized,type:attribute.array.constructor.name,sha256:sha(bytesOf(attribute))});

function decodeGlb(bytes) {
  assert.equal(bytes.readUInt32LE(0),0x46546c67,'Binary glTF required');
  assert.equal(bytes.readUInt32LE(4),2);
  assert.equal(bytes.readUInt32LE(8),bytes.length);
  const jsonLength=bytes.readUInt32LE(12), binHeader=20+jsonLength;
  assert.equal(bytes.readUInt32LE(16),0x4e4f534a);
  assert.equal(bytes.readUInt32LE(binHeader+4),0x004e4942);
  const binLength=bytes.readUInt32LE(binHeader);
  assert.equal(binHeader+8+binLength,bytes.length,'One JSON and one BIN chunk');
  const json=JSON.parse(bytes.subarray(20,binHeader).toString());
  assert.equal(json.buffers.length,1);
  assert.equal(json.buffers[0].uri,undefined);
  assert.ok(json.buffers[0].byteLength<=binLength);
  assert.ok(json.meshes.every(mesh=>mesh.primitives.length===1),'One material primitive per retained mesh');
  return {json,bin:bytes.subarray(binHeader+8)};
}
async function load(bytes) {
  const gltf=await new GLTFLoader().parseAsync(bytes.buffer.slice(bytes.byteOffset,bytes.byteOffset+bytes.byteLength),'');
  const meshes=[];gltf.scene.traverse(mesh=>{if(mesh.isMesh)meshes.push(mesh)});
  return meshes;
}
function encodeGlb(json, bin) {
  const text=Buffer.from(JSON.stringify(json));
  const padded=Buffer.alloc((text.length+3)&~3,0x20);text.copy(padded);
  assert.equal(bin.length%4,0);
  const header=Buffer.alloc(20), binHeader=Buffer.alloc(8);
  header.writeUInt32LE(0x46546c67,0);header.writeUInt32LE(2,4);header.writeUInt32LE(28+padded.length+bin.length,8);
  header.writeUInt32LE(padded.length,12);header.writeUInt32LE(0x4e4f534a,16);
  binHeader.writeUInt32LE(bin.length,0);binHeader.writeUInt32LE(0x004e4942,4);
  return Buffer.concat([header,padded,binHeader,bin]);
}

/** Transport only the native field. Original buffers, accessors, indices,
 * surfaces and scene structure stay exact; raw export UV echoes are omitted. */
export async function transportCanopyVisibility(baseBytes,nativeBytes) {
  const base=decodeGlb(baseBytes),native=decodeGlb(nativeBytes);
  for(const key of ['materials','nodes','scenes','scene','animations','skins','textures','images','samplers'])
    assert.deepEqual(native.json[key],base.json[key],'Original native '+key+' semantics');
  const before=await load(baseBytes),after=await load(nativeBytes);
  assert.equal(before.length,after.length);
  assert.equal(before.length,base.json.meshes.length);
  const additions=[];
  for(let meshIndex=0;meshIndex<before.length;meshIndex++) {
    const old=before[meshIndex],next=after[meshIndex],a=old.geometry,b=next.geometry;
    assert.equal(old.material.name,next.material.name);
    const keys=Object.keys(a.attributes).sort();
    for(const key of keys) {
      const x=field(a.attributes[key]),y=field(b.attributes[key]);
      delete x.sha256;delete y.sha256;assert.deepEqual(y,x,'Original vertex schema');
    }
    const tuple=(geometry,i)=>keys.map(key=>{
      const attr=geometry.attributes[key],size=attr.array.BYTES_PER_ELEMENT*attr.itemSize;
      return Buffer.from(attr.array.buffer,attr.array.byteOffset+i*size,size).toString('hex');
    }).join('/');
    const lookup=new Map();
    for(let i=0;i<a.attributes.position.count;i++) {
      const key=tuple(a,i);if(!lookup.has(key))lookup.set(key,[]);lookup.get(key).push(i);
    }
    const remap=[];
    for(let i=0;i<b.attributes.position.count;i++) {
      const matches=lookup.get(tuple(b,i));assert.ok(matches?.length,'Native vertex must have byte-exact original tuple');remap[i]=matches.pop();
    }
    assert.ok([...lookup.values()].every(matches=>matches.length===0),'Original vertex bijection');
    assert.equal(a.index.count,b.index.count);
    for(let i=0;i<a.index.count;i++)assert.equal(remap[b.index.getX(i)],a.index.getX(i),'Mapped triangle index sequence must stay exact');
    const added=Object.keys(b.attributes).filter(key=>!a.attributes[key]).sort();
    if(!old.material.name.startsWith('foliage')) { assert.deepEqual(added,[]);continue; }
    assert.deepEqual(added,['_canopy_visibility','uv1']);
    const raw=b.attributes._canopy_visibility,echo=b.attributes.uv1;
    assert.equal(raw.itemSize,2);assert.equal(raw.count,a.attributes.position.count);
    assert.equal(raw.normalized,false);assert.ok(raw.array instanceof Float32Array);
    assert.equal(echo.count,raw.count);assert.equal(echo.itemSize,2);
    const values=new Float32Array(raw.count*2);
    for(let i=0;i<raw.count;i++) {
      const x=raw.getX(i),y=raw.getY(i);
      assert.ok(Number.isFinite(x)&&x>=0&&x<=1&&Number.isFinite(y)&&y>=0&&y<=1,'Raw finite accessibility');
      assert.equal(echo.getX(i),x,'Redundant native UV echo x');
      assert.equal(echo.getY(i),Math.fround(1-y),'Redundant native UV echo y');
      values[remap[i]*2]=x;values[remap[i]*2+1]=y;
    }
    additions.push({meshIndex,material:old.material.name,values});
  }
  assert.equal(additions.length,2,'Both original oak foliage materials');
  const json=structuredClone(base.json),bins=[base.bin];let offset=base.bin.length;
  for(const row of additions) {
    const bytes=Buffer.from(row.values.buffer,row.values.byteOffset,row.values.byteLength);
    const view=json.bufferViews.push({buffer:0,byteOffset:offset,byteLength:bytes.length,target:34962})-1;
    const accessor=json.accessors.push({bufferView:view,componentType:5126,count:row.values.length/2,type:'VEC2'})-1;
    assert.equal(json.meshes[row.meshIndex].primitives[0].attributes._CANOPY_VISIBILITY,undefined);
    json.meshes[row.meshIndex].primitives[0].attributes._CANOPY_VISIBILITY=accessor;
    bins.push(bytes);offset+=bytes.length;
  }
  json.buffers[0].byteLength=offset;
  const bytes=encodeGlb(json,Buffer.concat(bins));
  const final=await load(bytes);
  for(let i=0;i<before.length;i++) {
    const a=before[i].geometry,b=final[i].geometry,row=additions.find(row=>row.meshIndex===i);
    assert.deepEqual(field(b.index),field(a.index));
    for(const key of Object.keys(a.attributes))assert.deepEqual(field(b.attributes[key]),field(a.attributes[key]),'Published original field remains byte-exact');
    assert.deepEqual(Object.keys(b.attributes).filter(key=>!a.attributes[key]),row?['_canopy_visibility']:[]);
    if(row)assert.equal(sha(bytesOf(b.attributes._canopy_visibility)),sha(Buffer.from(row.values.buffer)));
  }
  assert.ok(decodeGlb(bytes).bin.subarray(0,base.bin.length).equals(base.bin),'Original full BIN chunk remains byte-exact');
  return {bytes,verification:{originalGeometryPreserved:true,originalVertexOrderPreserved:true,originalSceneAndMaterialsPreserved:true,nativeUvEchoOmitted:true,
    fields:additions.map(row=>({material:row.material,vertices:row.values.length/2,bytes:row.values.byteLength,sha256:sha(Buffer.from(row.values.buffer))}))}};
}

/** Called while build-assets still owns unpublished staging. Every declared
 * base/native input is checked before the ordinary atomic publication phase. */
export async function completeCanopy({root,staging,models,generator}) {
  const directory=path.join(staging,'assets/source/canopy');
  const native=JSON.parse(readFileSync(path.join(directory,'native.json')));
  const base=JSON.parse(readFileSync(path.join(root,'assets/source/canopy-base/base.json')));
  assert.equal(base.version,1);assert.equal(generator.version,base.blenderVersion);
  const dependencies=['tools/assets/canopy-visibility.mjs',...Object.keys(native.inputs)];
  for(const [file,expected]of Object.entries(native.inputs)) {
    const resolved=path.resolve(root,file);assert.ok(resolved.startsWith(path.resolve(root)+path.sep));
    assert.equal(identify(resolved).sha256,expected,'Canopy input changed during native bake: '+file);
  }
  const records={};
  for(const name of models) {
    assert.ok(['tree','roadside-oak'].includes(name));
    const row=native.rows.find(row=>row.model===name);assert.ok(row);
    assert.equal(row.samples,256);
    const declared=base.models[name];
    for(const kind of ['blend','glb']) {
      assert.equal(declared[kind].path,`assets/source/canopy-base/${name}.${kind}`);
      const {path:file,...expected}=declared[kind];assert.deepEqual(identify(path.join(root,file)),expected);
    }
    const rawFile=path.join(directory,name+'.glb'),blendFile=path.join(directory,name+'.blend');
    assert.equal(identify(rawFile).sha256,row.glbSha256);assert.equal(identify(blendFile).sha256,row.sourceSha256);
    const transported=await transportCanopyVisibility(readFileSync(path.join(root,declared.glb.path)),readFileSync(rawFile));
    const output=path.join(staging,'apps/browser/public/assets');mkdirSync(output,{recursive:true});
    writeFileSync(path.join(output,name+'.glb'),transported.bytes);
    copyFileSync(blendFile,path.join(staging,'assets/source',name+'.blend'));
    const record={version:1,method:'Raw256-ray alpha-aware same-specimen isotropic diffuse accessibility; forward/back authored-normal hemispheres; no gain or clamp',
      base:declared,inputs:Object.fromEntries(dependencies.map(file=>[file,identify(path.join(root,file))])),
      native:{...row,blender:generator,rawExport:{path:`assets/source/canopy/${name}.glb`,...identify(rawFile)},source:{path:`assets/source/${name}.blend`,...identify(blendFile)}},
      ...transported.verification,output:{path:`apps/browser/public/assets/${name}.glb`,bytes:transported.bytes.length,sha256:sha(transported.bytes)}};
    writeFileSync(path.join(directory,name+'.json'),JSON.stringify(record,null,2)+'\n');records[name]=record;
  }
  return records;
}
